libfoundation/src/foundation-number.cpp

```diff
@@ -216,41 +304,1102 @@ bool __MCNumberCopyDescription(__MCNumber *self, MCStringRef& r_string)
+    {
+        return x - y;
+    }
+    
+    static inline bool signed_by_signed(integer_t x, integer_t y, integer_t& r_value)
+    {
+        return __checked_signed_subtract(x, y, &r_value);
+    }
+    
+    static inline bool unsigned_by_unsigned(uinteger_t x, uinteger_t y, uinteger_t& r_value)
+    {
+        return __checked_unsigned_subtract(x, y, &r_value);
+    }
+    
+    static inline bool signed_by_unsigned(integer_t x, uinteger_t y, integer_t& r_value)
+    {
+        // As we are guaranteed that y will be in the range (INT_MAX, UINT_MAX]
+        // if x is negative, we cannot produce an integer_t
+        if (x < 0)
+            return false;
+        
+        // x - y === -(y - x)
+        uinteger_t z;
+        if (!__checked_unsigned_subtract(y, x, &z))
+            return false;
+        
+        if (z > -INT32_MIN)
+            return false;
+        
+        r_value = -z;
+        
+        return true;
+    }
+    
+    static inline bool unsigned_by_signed(uinteger_t x, integer_t y, integer_t& r_value)
+    {
+        // As we are guaranteed that x will be in the range (INT_MAX, UINT_MAX]
+        // if y is negative, we cannot produce an integer_t
+        if (y < 0)
+            return false;
+        
+        // y is now known to be non-negative.
+        uinteger_t z;
+        if (!__checked_unsigned_subtract(x, (uinteger_t)y, &z))
+            return false;
+        
+        if (z > INT32_MAX)
+            return false;
+        
+        r_value = (integer_t)z;
+        
+        return true;
+    }
+};
+
+struct __MCNumberOperationMultiply
+{
+    static inline double double_by_double(double x, double y)
+    {
+        return x * y;
+    }
+    
+    static inline bigint_t bigint_by_bigint(bigint_t x, bigint_t y)
+    {
+        return x * y;
+    }
+    
+    static inline bool signed_by_signed(integer_t x, integer_t y, integer_t& r_value)
+    {
+        return __checked_signed_multiply(x, y, &r_value);
+    }
+    
+    static inline bool unsigned_by_unsigned(uinteger_t x, uinteger_t y, uinteger_t& r_value)
+    {
+        return __checked_unsigned_multiply(x, y, &r_value);
+    }
+    
+    static inline bool signed_by_unsigned(integer_t x, uinteger_t y, integer_t& r_value)
+    {
+        // We are guaranteed that y is in the range (INT_MAX, UINT_MAX] so
+        // x * y will only fit into an integer_t if x is 0.
+        if (x == 0)
+        {
+            r_value = 0;
+            return true;
+        }
+        
+        return false;
+    }
+    
+    static inline bool unsigned_by_signed(uinteger_t x, integer_t y, integer_t& r_value)
+    {
+        // We are guaranteed that y is in the range (INT_MAX, UINT_MAX] so
+        // x * y will only fit into an integer_t if y is 0.
+        if (y == 0)
+        {
+            r_value = 0;
+            return true;
+        }
+        
+        return false;
+    }
+};
+
+// The '/' operator always produces reals (use div or mod for integer ops).
```

Comment:
The problem with that is that you don't know what the container type is (polymorphic selection is based on the dynamic type of inputs - inout arguments don't help here since the 'in' part since a type can be passed in to any of its supertypes).

All the other arithmetic operations have the property that the integer forms are 'subsets' (in some sense) of the real forms... Division is different as integer division loses information.